sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/SdkTracerProviderBuilder.scala

```diff
@@ -0,0 +1,155 @@
+package trace
+
+import cats.effect.Async
+import cats.syntax.functor._
+import org.typelevel.otel4s.sdk.{Resource => InstrumentResource}
+import org.typelevel.otel4s.sdk.context.LocalContext
+import org.typelevel.otel4s.sdk.context.propagation.ContextPropagators
+import org.typelevel.otel4s.sdk.context.propagation.TextMapPropagator
+import org.typelevel.otel4s.sdk.trace.samplers.Sampler
+
+/** Builder for [[SdkTracerProvider]].
+  */
+sealed trait SdkTracerProviderBuilder[F[_]] {
+
+  /** Sets an [[IdGenerator]].
+    *
+    * [[IdGenerator]] will be used each time a
+    * [[org.typelevel.otel4s.trace.Span Span]] is started.
+    *
+    * ''Note'': the id generator must be thread-safe and return immediately (no
+    * remote calls, as contention free as possible).
+    */
+  def setIdGenerator(idGenerator: IdGenerator[F]): SdkTracerProviderBuilder[F]
+
+  /** Sets a [[org.typelevel.otel4s.sdk.Resource]] to be attached to all spans
+    * created by [[org.typelevel.otel4s.trace.Tracer Tracer]].
+    */
+  def setResource(resource: InstrumentResource): SdkTracerProviderBuilder[F]
+
+  /** Merges a [[org.typelevel.otel4s.sdk.Resource]] with the current one.
+    */
+  def addResource(resource: InstrumentResource): SdkTracerProviderBuilder[F]
+
+  /** Sets an initial [[SpanLimits]] that should be used with this SDK.
+    *
+    * The limits will be used for every
+    * [[org.typelevel.otel4s.trace.Span Span]].
+    */
+  def setSpanLimits(limits: SpanLimits): SdkTracerProviderBuilder[F]
+
+  /** Sets a [[org.typelevel.otel4s.sdk.trace.samplers.Sampler Sampler]] to use
+    * for sampling traces.
+    *
+    * Sampler will be called each time a
+    * [[org.typelevel.otel4s.trace.Span Span]] is started.
+    *
+    * ''Note:'' the sampler must be thread-safe and return immediately (no
+    * remote calls, as contention free as possible).
+    */
+  def setSampler(sampler: Sampler): SdkTracerProviderBuilder[F]
+
+  /** Adds a [[TextMapPropagator]]s to use for the context propagation.
+    */
+  def addTextMapPropagators(
+      propagators: TextMapPropagator*
+  ): SdkTracerProviderBuilder[F]
+
+  /** Adds a [[SpanProcessor]] to the span processing pipeline that will be
+    * built.
+    *
+    * [[SpanProcessor]] will be called each time a
+    * [[org.typelevel.otel4s.trace.Span Span]] is started or ended.
+    *
+    * ''Note:'' the span processor must be thread-safe and return immediately
+    * (no remote calls, as contention free as possible).
+    */
+  def addSpanProcessor(
+      processor: SpanProcessor[F]
+  ): SdkTracerProviderBuilder[F]
+
+  /** Creates a new [[SdkTracerProvider]] with configuration of this builder.
+    */
+  def build: F[SdkTracerProvider[F]]
+}
+
+object SdkTracerProviderBuilder {
+
+  /** Creates a new [[SdkTracerProviderBuilder]] with default configuration.
+    */
+  def default[F[_]: Async: LocalContext]: SdkTracerProviderBuilder[F] =
+    new Builder[F](
+      idGeneratorF = IdGenerator.default[F],
```

Comment:
`IdGenerator.default[F]` requires `Sync` (to use  `cats.effect.std.Random.scalaUtilRandom`) and the `SdkTraceProvider` requires `Concurrent : Clock` (or `Temporal`).

Would it be better to change effect constraints to:
```
def default[F[_]: Temporal: Random: LocalContext]: SdkTracerProviderBuilder[F]
```

Reply:
Yes, that would be better 👍 in general, I am skeptical if `Sync` or `Async` constraints appear outside of CE/FS2.

Side note, we could implement our own `Random` in CE that only needs `Concurrent` and uses `Ref`.